### PR TITLE
[State Exporter] support pre-checkpointing reporter

### LINKED ISSUE #2342: Refactor checkpoint export to support the execution of reports before or in parallel of checkpoint writing
## Problem Definition

We would like to refactor the [ExportCheckpointAt](https://github.com/onflow/flow-go/blob/d99292e8774132b680a00cf136ef633e1303b725/ledger/complete/ledger.go#L267) to more cleanly support the execution of reports before the writing of the root.checkpoint file, or we would like to add support for executing the writing of the root.checkpoint in parallel with the execution of the reporters.

## Definition of Done

* The `ExportCheckpointAt` method is refactored to allow reports to be executed in parallel or a pattern is made available for executing reports before the root checkpoint is written to disk.
* Before changes are merged and this ticket is closed, the util should be built and tested by executing a state extraction

### PR DESCRIPTION
**Pull request by ramtinms**
closes #2342

Improving the code by adding pre-checkpointer reporters, 
Why? 
- for faster sporks, we need a way to export some info as soon as possible before generating the checkpoint
- some reports are time consuming and should be done after the checkpoint generation
- iteration over maps are not deterministic

### CODE REVIEW

File: ledger/complete/ledger.go [hunk 1]
```diff
@@ -356,17 +356,19 @@ func (l *Ledger) ExportCheckpointAt(
...
 	l.logger.Info().Msgf("successfully built new trie. NEW ROOT STATECOMMIEMENT: %v", statecommitment.String())
 
-	// If defined, run extraction report BEFORE writing checkpoint file
-	// This is used to optmize the spork process
-	if extractionReport, ok := reporters[extractionReportName]; ok {
-		err := runReport(extractionReport, payloads, l.logger)
+	l.logger.Info().Msgf("running pre-checkpoint reporters")
+	// run post migration reporters
+	for _, reporter := range preCheckpointReporters {
+		l.logger.Info().Msgf("running a pre-checkpoint generation reporter: %s", reporter.Name())
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:362]:
```suggestion
	for i, reporter := range preCheckpointReporters {
		l.logger.Info().Msgf("running a pre-checkpoint generation reporter: %s, (%v/%v)", reporter.Name(), i, len(preCheckpointReporters))
```

File: ledger/complete/ledger.go [hunk 2]
```diff
@@ -390,17 +392,18 @@ func (l *Ledger) ExportCheckpointAt(
...
 	l.logger.Info().Msgf("checkpoint file successfully stored at: %v %v", outputDir, outputFile)
 
-	l.logger.Info().Msgf("generating reports")
+	l.logger.Info().Msgf("finished running post-checkpoint reporters")
 
-	// run reporters
-	for _, reporter := range reporters {
+	// running post checkpoint reporters
+	for _, reporter := range postCheckpointReporters {
+		l.logger.Info().Msgf("running a post-checkpoint generation reporter: %s", reporter.Name())
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:399]:
```suggestion
	for i, reporter := range postCheckpointReporters {
		l.logger.Info().Msgf("running a post-checkpoint generation reporter: %s, (%v/%v)", reporter.Name(), (i, len(postCheckpointReporters))
```

File: ledger/complete/ledger.go [hunk 3]
```diff
@@ -390,17 +392,18 @@ func (l *Ledger) ExportCheckpointAt(
...
 		err := runReport(reporter, payloads, l.logger)
 		if err != nil {
 			return ledger.State(hash.DummyHash), err
 		}
 	}
 
-	l.logger.Info().Msgf("all reports genereated")
+	l.logger.Info().Msgf("ran all pre-checkpoint reporters")
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:406]:
```suggestion
	l.logger.Info().Msgf("ran all post-checkpoint reporters")
```

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go [hunk 4]
```diff
@@ -86,28 +85,29 @@ func extractExecutionState(
 		log.Info().Msgf("preparing reporter files")
 		reportFileWriterFactory := reporters.NewReportFileWriterFactory(outputDir, log)
 
-		rs = map[string]ledger.Reporter{
-			// The ExportReporter needs to be run first so that it can be used
-			// immediately after execution
-			extractionReportName: reporters.NewExportReporter(log,
+		preCheckpointReporters = []ledger.Reporter{
+			reporters.NewExportReporter(log,
```
Comment (zhangchiqing) [inline cmd/util/cmd/execution-state-extract/execution_state_extract.go:89]:
```suggestion
			// report epoch counter which is needed for finalizing root block
			reporters.NewExportReporter(log,
```

**Review by zhangchiqing — APPROVED** [review]
LGTM

**Comment by sjonpaulbrown** [comment]
To be safe, I will build the util and execute state extraction to ensure that these changes work as expected. The code changes LGTM. I will follow up with the results.

File: ledger/complete/ledger.go [hunk 5]
```diff
@@ -409,17 +411,18 @@ func (l *Ledger) ExportCheckpointAt(
...
 
-	l.logger.Info().Msgf("generating reports")
+	l.logger.Info().Msgf("finished running post-checkpoint reporters")
 
-	// run reporters
-	for _, reporter := range reporters {
-		err := runReport(reporter, payloads, statecommitment, l.logger)
+	// running post checkpoint reporters
+	for i, reporter := range postCheckpointReporters {
+		l.logger.Info().Msgf("running a post-checkpoint generation reporter: %s, (%v/%v)", reporter.Name(), (i, len(postCheckpointReporters))
```
Comment (sjonpaulbrown) [inline ledger/complete/ledger.go:418]:
```suggestion
		l.logger.Info().Msgf("running a post-checkpoint generation reporter: %s, (%v/%v)", reporter.Name(), i, len(postCheckpointReporters))
```

This is causing build errors.

**Comment by sjonpaulbrown** [comment]
After fixing the formatting error, everything built and ran as expected. I validated that the pre-checkpoint and post-checkpoint reporters ran as expected during state extraction.

**Review by sjonpaulbrown — APPROVED** [review]

**Review by zhangchiqing — APPROVED** [review]

**Pull request merged**